http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java

```diff
@@ -169,28 +173,42 @@ private void releaseParentChannel(Channel parentChannel) {
 
     @Override
     public void close() {
-        try {
-            setClosedFlag().await();
-            for (MultiplexedChannelRecord c : connections) {
-                Future<Channel> f = c.getConnectionFuture();
-                f.await();
-                if (f.isSuccess()) {
-                    connectionPool.release(f.getNow()).await();
-                }
-            }
+        doInEventLoop(eventLoop, this::close0);
+    }
+
+    private void close0() {
+        if (closed) {
+            return;
+        }
+
+        closed = true;
+
+        if (connections.isEmpty()) {
             connectionPool.close();
-        } catch (InterruptedException ie) {
-            Thread.currentThread().interrupt();
-            throw new RuntimeException(ie);
+            return;
         }
-    }
 
-    private Promise<Void> setClosedFlag() {
-        Promise<Void> closedFuture = eventLoop.newPromise();
-        doInEventLoop(eventLoop, () -> {
-            closed = true;
-            closedFuture.setSuccess(null);
-        });
-        return closedFuture;
+        AtomicLong unreleasedConnections = new AtomicLong(connections.size());
+
+        for (MultiplexedChannelRecord record : connections) {
+            record.getConnectionFuture().addListener((Future<Channel> connectionFuture) -> {
+                if (connectionFuture.isSuccess()) {
+                    Channel connection = connectionFuture.getNow();
+                    connectionPool.release(connection).addListener((Future<Void> releaseFuture) -> {
+                        if (!releaseFuture.isSuccess()) {
```

Comment:
See above comment re: closing the channel

Reply:
Oh actually, nm I didn't read the javadoc carefully. So It looks like we don't need to worry about closing the channel beforehand since either it will get released and the closed when we close the pool, or the release will fail and we it gets closed anyway.